src/upgrades/upgrade.py

```diff
@@ -231,3 +250,47 @@ def upgrade_unit(self, *, charm) -> None:
 
         Only applies to machine charm
         """
+
+    def pre_upgrade_check(self) -> None:
+        """Check if this app is ready to upgrade.
+
+        Runs before any units are upgraded
+
+        Does *not* run during rollback
+
+        On machines, this runs before any units are upgraded (after `juju refresh`)
+        On machines & Kubernetes, this also runs during pre-upgrade-check action
+
+        Can run on leader or non-leader unit
+
+        Raises:
+            PrecheckFailed: App is not ready to upgrade
+
+        TODO Kubernetes: Run (some) checks after `juju refresh` (in case user forgets to run
+        pre-upgrade-check action). Note: 1 unit will upgrade before we can run checks (checks may
+        need to be modified).
+        See https://chat.canonical.com/canonical/pl/cmf6uhm1rp8b7k8gkjkdsj4mya
+        """
+        logger.debug("Running pre-upgrade checks")
+
+        # TODO In future PR when we support upgrades on sharded clusters, have the shard verify
+        # that the config-server has already upgraded.
+
+        try:
+            self._charm.upgrade.wait_for_cluster_healthy()
+        except RetryError:
+            logger.error("Cluster is not healthy")
+            raise PrecheckFailed("Cluster is not healthy")
+
+        # We do not get to decide the order of units to upgrade, so we move the primary to the
```

Comment:
I think we do get to decide the order of the units to upgrade, no?

Reply:
No we dont. In V1 implementation we did, but in V2, it is highest to lowest

Reply:
Wait I am confused. So @carlcsaposs-canonical we **can** choose the order?

Reply:
If so, how can we do that. It would alleviate a lot of pain because then we could remove the "MovePrimary" logic

Reply:
My apologise Carl, I had forgetten and of course yu are correct. We **can** choose the order, but **only on VMs**

Back in April MongoDB team discussed the differences in K8s + VM implementations

And although we dislike the movePrimary functionality on VM (since we could choose the order of upgrades anyway) - we had decided that the benefits of keeping the charms insync out-weighed the benefits. So we decided that both charms should maintain the upgrade order of highest to lowest.

Reply:
Let me update the comment to reflect this.